File: services/catalog/internal/events/publisher.go
package events

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	amqp "github.com/rabbitmq/amqp091-go"
	"go.uber.org/zap"
)

const (
	exchangeName = "bookstore.events"
	exchangeType = "topic"

	// Event types
	EventTypeCatalogCreated = "catalog.created"
	EventTypeCatalogUpdated = "catalog.updated"
	EventTypeCatalogDeleted = "catalog.deleted"

	// Retry configuration
	maxRetries     = 3
	initialBackoff = 100 * time.Millisecond
	maxBackoff     = 5 * time.Second
)

// Publisher handles event publishing to RabbitMQ
type Publisher struct {
	conn    *amqp.Connection
	channel *amqp.Channel
	log     *zap.Logger
}

// Event represents a domain event
type Event struct {
	EventID       string                 `json:"event_id"`
	EventType     string                 `json:"event_type"`
	EventVersion  string                 `json:"event_version"`
	Timestamp     string                 `json:"timestamp"`
	CorrelationID string                 `json:"correlation_id,omitempty"`
	Payload       map[string]interface{} `json:"payload"`
}

// NewPublisher creates a new event publisher
func NewPublisher(url string, log *zap.Logger) (*Publisher, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RabbitMQ: %w", err)
	}

	channel, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, fmt.Errorf("failed to open channel: %w", err)
	}

	// Declare exchange
	if err := channel.ExchangeDeclare(
		exchangeName,
		exchangeType,
		true,  // durable
		false, // auto-deleted
		false, // internal
		false, // no-wait
		nil,   // arguments
	); err != nil {
		channel.Close()
		conn.Close()
		return nil, fmt.Errorf("failed to declare exchange: %w", err)
	}

	// Enable publisher confirms for reliability
	if err := channel.Confirm(false); err != nil {
		channel.Close()
		conn.Close()
		return nil, fmt.Errorf("failed to enable publisher confirms: %w", err)
	}

	log.Info("Connected to RabbitMQ", zap.String("exchange", exchangeName))

	return &Publisher{
		conn:    conn,
		channel: channel,
		log:     log,
	}, nil
}

// PublishBookCreated publishes a book created event
func (p *Publisher) PublishBookCreated(ctx context.Context, sku, title, author, category, currency string, price int64, active bool) error {
	event := Event{
		EventID:      uuid.New().String(),
		EventType:    EventTypeCatalogCreated,
		EventVersion: "1.0.0",
		Timestamp:    time.Now().UTC().Format(time.RFC3339),
		Payload: map[string]interface{}{
			"sku":      sku,
			"title":    title,
			"author":   author,
			"price":    price,
			"currency": currency,
			"category": category,
			"active":   active,
		},
	}

	// Extract correlation ID from context if available
	if corrID := ctx.Value("correlation_id"); corrID != nil {
		event.CorrelationID = corrID.(string)
	}

	return p.publishWithRetry(ctx, EventTypeCatalogCreated, event)
}

// PublishBookUpdated publishes a book updated event
func (p *Publisher) PublishBookUpdated(ctx context.Context, sku string, fieldsChanged []string, updates map[string]interface{}) error {
	event := Event{
		EventID:      uuid.New().String(),
		EventType:    EventTypeCatalogUpdated,
		EventVersion: "1.0.0",
		Timestamp:    time.Now().UTC().Format(time.RFC3339),
		Payload: map[string]interface{}{
			"sku":            sku,
			"fields_changed": fieldsChanged,
		},
	}

	// Add updated field values to payload
	for k, v := range updates {
		event.Payload[k] = v
	}

	// Extract correlation ID from context if available
	if corrID := ctx.Value("correlation_id"); corrID != nil {
		event.CorrelationID = corrID.(string)
	}

	return p.publishWithRetry(ctx, EventTypeCatalogUpdated, event)
}

// PublishBookDeleted publishes a book deleted event
func (p *Publisher) PublishBookDeleted(ctx context.Context, sku string) error {
	event := Event{
		EventID:      uuid.New().String(),
		EventType:    EventTypeCatalogDeleted,
		EventVersion: "1.0.0",
		Timestamp:    time.Now().UTC().Format(time.RFC3339),
		Payload: map[string]interface{}{
			"sku": sku,
		},
	}

	// Extract correlation ID from context if available
	if corrID := ctx.Value("correlation_id"); corrID != nil {
		event.CorrelationID = corrID.(string)
	}

	return p.publishWithRetry(ctx, EventTypeCatalogDeleted, event)
}

// publishWithRetry publishes an event with exponential backoff retry
func (p *Publisher) publishWithRetry(ctx context.Context, routingKey string, event Event) error {
	body, err := json.Marshal(event)
	if err != nil {
		p.log.Error("Failed to marshal event", zap.Error(err))
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	backoff := initialBackoff
	var lastErr error

	for attempt := 0; attempt < maxRetries; attempt++ {
		if attempt > 0 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(backoff):
				backoff *= 2
				if backoff > maxBackoff {
					backoff = maxBackoff
				}
			}
		}

		// Publish with confirmation
		confirms := p.channel.NotifyPublish(make(chan amqp.Confirmation, 1))

		err := p.channel.PublishWithContext(
			ctx,
			exchangeName,
			routingKey,
			false, // mandatory
			false, // immediate
			amqp.Publishing{
				ContentType:  "application/json",
				DeliveryMode: amqp.Persistent,
				Timestamp:    time.Now(),
				MessageId:    event.EventID,
				Body:         body,
				Headers: amqp.Table{
					"event_type":    event.EventType,
					"event_version": event.EventVersion,
				},
			},
		)

		if err != nil {
			lastErr = err
			p.log.Warn("Failed to publish event, retrying",
				zap.Int("attempt", attempt+1),
				zap.Error(err),
			)
			continue
		}

		// Wait for confirmation
		select {
		case confirm := <-confirms:
			if confirm.Ack {
				p.log.Info("Event published successfully",
					zap.String("event_id", event.EventID),
					zap.String("event_type", event.EventType),
					zap.String("routing_key", routingKey),
				)
				return nil
			}
			lastErr = fmt.Errorf("event not acknowledged")
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(5 * time.Second):
			lastErr = fmt.Errorf("confirmation timeout")
		}

		p.log.Warn("Event publish not confirmed, retrying",
			zap.Int("attempt", attempt+1),
			zap.Error(lastErr),
		)
	}

	p.log.Error("Failed to publish event after retries",
		zap.String("event_id", event.EventID),
		zap.String("event_type", event.EventType),
		zap.Int("attempts", maxRetries),
		zap.Error(lastErr),
	)
	return fmt.Errorf("failed to publish event after %d attempts: %w", maxRetries, lastErr)
}

// IsHealthy checks if the publisher connection is healthy
func (p *Publisher) IsHealthy() bool {
	return p.conn != nil && !p.conn.IsClosed()
}

// Close closes the publisher connection
func (p *Publisher) Close() error {
	if p.channel != nil {
		if err := p.channel.Close(); err != nil {
			p.log.Error("Failed to close channel", zap.Error(err))
		}
	}
	if p.conn != nil {
		if err := p.conn.Close(); err != nil {
			p.log.Error("Failed to close connection", zap.Error(err))
			return err
		}
	}
	p.log.Info("Publisher closed")
	return nil
}

File: services/cart/cart/events/publisher.py
"""
RabbitMQ event publisher with retry logic and confirmations
"""
import json
import time
import uuid
from datetime import datetime
from typing import Dict, Any, Optional
import pika
from pika.exceptions import AMQPConnectionError, AMQPChannelError
import structlog

from cart.config import Config


logger = structlog.get_logger()


class EventPublisher:
    """Publisher for cart domain events to RabbitMQ"""
    
    def __init__(self, config: Config):
        self.config = config
        self.connection: Optional[pika.BlockingConnection] = None
        self.channel: Optional[pika.channel.Channel] = None
        self.logger = structlog.get_logger().bind(component="event_publisher")
        self._connect()
    
    def _connect(self) -> None:
        """Establish connection to RabbitMQ"""
        try:
            parameters = pika.URLParameters(self.config.rabbitmq_url)
            parameters.heartbeat = 30
            parameters.blocked_connection_timeout = 300
            
            self.connection = pika.BlockingConnection(parameters)
            self.channel = self.connection.channel()
            
            # Declare exchange
            self.channel.exchange_declare(
                exchange=self.config.rabbitmq_cart_exchange,
                exchange_type='topic',
                durable=True
            )
            
            # Enable publisher confirms
            self.channel.confirm_delivery()
            
            self.logger.info("Connected to RabbitMQ for publishing",
                           exchange=self.config.rabbitmq_cart_exchange)
            
        except AMQPConnectionError as e:
            self.logger.error("Failed to connect to RabbitMQ", error=str(e))
            raise
    
    def _reconnect(self) -> None:
        """Reconnect to RabbitMQ"""
        try:
            self.close()
        except:
            pass
        self._connect()
    
    def _publish_with_retry(
        self,
        routing_key: str,
        event: Dict[str, Any]
    ) -> None:
        """
        Publish event with exponential backoff retry
        
        Args:
            routing_key: RabbitMQ routing key
            event: Event payload
            
        Raises:
            Exception if all retries fail
        """
        start_time = time.time()
        
        event_json = json.dumps(event)
        max_attempts = self.config.rabbitmq_retry_max_attempts
        delay = self.config.rabbitmq_retry_initial_delay
        
        for attempt in range(max_attempts):
            try:
                # Ensure connection is alive
                if not self.is_healthy():
                    self._reconnect()
                
                # Publish with confirmation
                self.channel.basic_publish(
                    exchange=self.config.rabbitmq_cart_exchange,
                    routing_key=routing_key,
                    body=event_json,
                    properties=pika.BasicProperties(
                        content_type='application/json',
                        delivery_mode=2,  # Persistent
                        message_id=event['event_id'],
                        timestamp=int(time.time()),
                        headers={
                            'event_type': event['event_type'],
                            'event_version': event.get('event_version', '1.0.0')
                        }
                    ),
                    mandatory=False
                )
                
                # Record success metrics
                duration = time.time() - start_time
                event_publish_duration.observe(duration)
                cart_events_published_total.labels(
                    event_type=event['event_type'],
                    status='success'
                ).inc()
                
                self.logger.info(
                    "Event published",
                    event_id=event['event_id'],
                    event_type=event['event_type'],
                    routing_key=routing_key,
                    attempt=attempt + 1
                )
                return
                
            except (AMQPConnectionError, AMQPChannelError) as e:
                self.logger.warning(
                    "Event publish failed, retrying",
                    event_id=event['event_id'],
                    attempt=attempt + 1,
                    max_attempts=max_attempts,
                    error=str(e)
                )
                
                if attempt < max_attempts - 1:
                    time.sleep(delay)
                    delay = min(delay * 2, self.config.rabbitmq_retry_max_delay)
                else:
                    # All retries failed
                    cart_events_published_total.labels(
                        event_type=event['event_type'],
                        status='failed'
                    ).inc()
                    self.logger.error(
                        "Event publish failed after all retries",
                        event_id=event['event_id'],
                        event_type=event['event_type']
                    )
                    raise
    
    def publish_item_added(
        self,
        user_id: str,
        sku: str,
        qty: int,
        price: float,
        currency: str,
        title: str,
        correlation_id: Optional[str] = None
    ) -> None:
        """Publish cart.item_added event"""
        event = {
            'event_id': str(uuid.uuid4()),
            'event_type': 'cart.item_added',
            'event_version': '1.0.0',
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'correlation_id': correlation_id,
            'payload': {
                'user_id': user_id,
                'sku': sku,
                'qty': qty,
                'price': int(price * 100),  # Convert to cents
                'currency': currency,
                'title': title
            }
        }
        
        self._publish_with_retry('cart.item_added', event)
    
    def publish_item_removed(
        self,
        user_id: str,
        sku: str,
        qty: int,
        correlation_id: Optional[str] = None
    ) -> None:
        """Publish cart.item_removed event"""
        event = {
            'event_id': str(uuid.uuid4()),
            'event_type': 'cart.item_removed',
            'event_version': '1.0.0',
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'correlation_id': correlation_id,
            'payload': {
                'user_id': user_id,
                'sku': sku,
                'qty': qty
            }
        }
        
        self._publish_with_retry('cart.item_removed', event)
    
    def publish_cart_cleared(
        self,
        user_id: str,
        reason: str = "user_requested",
        correlation_id: Optional[str] = None
    ) -> None:
        """Publish cart.cleared event"""
        event = {
            'event_id': str(uuid.uuid4()),
            'event_type': 'cart.cleared',
            'event_version': '1.0.0',
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'correlation_id': correlation_id,
            'payload': {
                'user_id': user_id,
                'reason': reason
            }
        }
        
        self._publish_with_retry('cart.cleared', event)
    
    def publish_checkout_requested(
        self,
        user_id: str,
        items: list,
        total_amount: float,
        currency: str,
        correlation_id: Optional[str] = None
    ) -> None:
        """Publish cart.checkout_requested event"""
        event = {
            'event_id': str(uuid.uuid4()),
            'event_type': 'cart.checkout_requested',
            'event_version': '1.0.0',
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'correlation_id': correlation_id,
            'payload': {
                'user_id': user_id,
                'items': items,
                'total_amount': int(total_amount * 100),
                'currency': currency
            }
        }
        
        self._publish_with_retry('cart.checkout_requested', event)
    
    def is_healthy(self) -> bool:
        """Check if publisher connection is healthy"""
        try:
            return (
                self.connection is not None and
                self.connection.is_open and
                self.channel is not None and
                self.channel.is_open
            )
        except:
            return False
    
    def close(self) -> None:
        """Close publisher connection"""
        try:
            if self.channel and self.channel.is_open:
                self.channel.close()
            if self.connection and self.connection.is_open:
                self.connection.close()
            self.logger.info("Event publisher closed")
        except Exception as e:
            self.logger.error("Error closing publisher", error=str(e))




